SingularityService/src/main/java/com/hubspot/singularity/mesos/SingularityMesosTaskBuilder.java

```diff
@@ -163,63 +166,76 @@ private boolean hasLiteralPortMapping(Optional<SingularityContainerInfo> maybeCo
+    envVars.put("TASK_HOST", offerHolder.getHostname());
 
-    setEnv(envBldr, "TASK_RACK_ID", offerHolder.getRackId());
+    envVars.put("TASK_RACK_ID", offerHolder.getRackId());
 
-    setEnv(envBldr, "TASK_REQUEST_ID", task.getPendingTask().getPendingTaskId().getRequestId());
-    setEnv(envBldr, "TASK_DEPLOY_ID", taskId.getDeployId());
-    setEnv(envBldr, "TASK_ID", taskId.getId());
-    setEnv(envBldr, "ESTIMATED_INSTANCE_COUNT", task.getRequest().getInstancesSafe());
+    envVars.put("TASK_REQUEST_ID", task.getPendingTask().getPendingTaskId().getRequestId());
+    envVars.put("TASK_DEPLOY_ID", taskId.getDeployId());
+    envVars.put("TASK_ID", taskId.getId());
+    envVars.put("ESTIMATED_INSTANCE_COUNT", task.getRequest().getInstancesSafe());
 
     if (task.getPendingTask().getUser().isPresent()) {
-      setEnv(envBldr, "STARTED_BY_USER", task.getPendingTask().getUser().get());
+      envVars.put("STARTED_BY_USER", task.getPendingTask().getUser().get());
     }
 
     for (Entry<String, String> envEntry : task.getDeploy().getEnv().or(Collections.<String, String>emptyMap()).entrySet()) {
-      setEnv(envBldr, envEntry.getKey(), fillInTaskIdValues(envEntry.getValue(), offerHolder, taskId));
+      envVars.put(envEntry.getKey(), fillInTaskIdValues(envEntry.getValue(), offerHolder, taskId));
     }
 
     if (task.getDeploy().getTaskEnv().isPresent() && task.getDeploy().getTaskEnv().get().containsKey(taskId.getInstanceNo()) && !task.getDeploy().getTaskEnv().get().get(taskId.getInstanceNo()).isEmpty()) {
       for (Entry<String, String> envEntry : task.getDeploy().getTaskEnv().get().get(taskId.getInstanceNo()).entrySet()) {
-        setEnv(envBldr, envEntry.getKey(), fillInTaskIdValues(envEntry.getValue(), offerHolder, taskId));
+        envVars.put(envEntry.getKey(), fillInTaskIdValues(envEntry.getValue(), offerHolder, taskId));
       }
     }
 
     if (ports.isPresent()) {
       for (int portNum = 0; portNum < ports.get().length; portNum++) {
         if (portNum == 0) {
-          setEnv(envBldr, "PORT", ports.get()[portNum]);
+          envVars.put("PORT", ports.get()[portNum]);
         }
 
-        setEnv(envBldr, String.format("PORT%s", portNum), ports.get()[portNum]);
+        envVars.put(String.format("PORT%s", portNum), ports.get()[portNum]);
       }
     }
 
     if (task.getPendingTask().getResources().isPresent()) {
       Resources override = task.getPendingTask().getResources().get();
 
       if (override.getCpus() != 0) {
-        setEnv(envBldr, "DEPLOY_CPUS", ((long) override.getCpus()));
+        envVars.put("DEPLOY_CPUS", ((long) override.getCpus()));
       }
 
       if (override.getMemoryMb() != 0) {
-        setEnv(envBldr, "DEPLOY_MEM", ((long) override.getMemoryMb()));
+        envVars.put("DEPLOY_MEM", ((long) override.getMemoryMb()));
       }
+    }
 
+    if (task.getDeploy().getRunImmediately().isPresent()) {
+      SingularityRunNowRequest runNowRequest = task.getDeploy().getRunImmediately().get();
```

Comment:
As mentioned below, this is a field that specifies config to run immediately upon successful deploy. The `SingularityPendingTask` is where the environment override needs to be trasferred to first. So, something like:

- New RunNowRequest comes in with environmentOverrides and the `SingularityRunNowRequest` is saved as part of the `SingularityPendingRequest` and put into the pending queue (triggered from the `RequestResource`
- overrides from RunNowRequest saved as part of the SingularityPendingTask when it is created based on that pending request from the queue (this is the piece that is currently missing)
- When the task builder is ready to build the task from the pending task, it can then read the environment overrides from the pending task and apply those here

Reply:
Next diff will have changes that I believe satisfy this, though I'm not sure my implementation is as clean as it could be. Either way, let me know and I'll work through it.